lib/Interpreter/CppInterOp.cpp

```diff
@@ -1027,61 +1056,78 @@ namespace Cpp {
   }
 
   TCppFunction_t
-  BestTemplateFunctionMatch(const std::vector<TCppFunction_t>& candidates,
+  BestOverloadFunctionMatch(const std::vector<TCppFunction_t>& candidates,
                             const std::vector<TemplateArgInfo>& explicit_types,
                             const std::vector<TemplateArgInfo>& arg_types) {
-
-    for (const auto& candidate : candidates) {
-      auto* TFD = (FunctionTemplateDecl*)candidate;
-      clang::TemplateParameterList* tpl = TFD->getTemplateParameters();
-
-      // template parameter size does not match
-      if (tpl->size() < explicit_types.size())
-        continue;
-
-      // right now uninstantiated functions give template typenames instead of
-      // actual types. We make this match solely based on count
-
-      const FunctionDecl* func = TFD->getTemplatedDecl();
+    auto& S = getSema();
+    auto& C = S.getASTContext();
 
 #ifdef CPPINTEROP_USE_CLING
-      if (func->getNumParams() > arg_types.size())
-        continue;
-#else // CLANG_REPL
-      if (func->getMinRequiredArguments() > arg_types.size())
-        continue;
+    cling::Interpreter::PushTransactionRAII RAII(&getInterp());
 #endif
 
-      // TODO(aaronj0) : first score based on the type similarity before forcing
-      // instantiation.
-
-      TCppFunction_t instantiated =
-          InstantiateTemplate(candidate, arg_types.data(), arg_types.size());
-      if (instantiated)
-        return instantiated;
-
-      // Force the instantiation with template params in case of no args
-      // maybe steer instantiation better with arg set returned from
-      // TemplateProxy?
-      instantiated = InstantiateTemplate(candidate, explicit_types.data(),
-                                          explicit_types.size());
-      if (instantiated)
-        return instantiated;
-
-      // join explicit and arg_types
-      std::vector<TemplateArgInfo> total_arg_set;
-      total_arg_set.reserve(explicit_types.size() + arg_types.size());
-      total_arg_set.insert(total_arg_set.end(), explicit_types.begin(),
-                           explicit_types.end());
-      total_arg_set.insert(total_arg_set.end(), arg_types.begin(),
-                           arg_types.end());
-
-      instantiated = InstantiateTemplate(candidate, total_arg_set.data(),
-                                         total_arg_set.size());
-      if (instantiated)
-        return instantiated;
+    llvm::SmallVector<Expr*> Args;
```

Comment:
Here we should write a comment of the sort:

"The overload resolution interfaces in Sema require a list of expressions. However, unlike handwritten C++, we do not always have a expression. Here we synthesize a placeholder expression to be able to use Sema::AddOverloadCandidate. Made up expressions are fine because the interface uses the list size and the expression types."